File: App/Resources/AlgorithmDetails/pairwisemergesortiterative/pairwisemergesortiterative.kt
fun compSwap(arr: Array<Int>, a: Int, b: Int, end: Int) {
  if (b < end && arr[a] > arr[b]) {
    arr[a] = arr[b].also { arr[b] = arr[a] }
  }
}

fun sort(arr: Array<Int>) {
  val length = arr.size
  val end = length

  var n = 1
  while (n < length) n = n shl 1

  var k = n shr 1
  while (k > 0) {
    var j = 0
    while (j < length) {
      for (i in 0 until k) {
        compSwap(arr, j + i, j + k + i, end)
      }
      j += k shl 1
    }
    k = k shr 1
  }

  k = 2
  while (k < n) {
    var m = k shr 1
    while (m > 0) {
      var j = 0
      while (j < length) {
        var p = m
        while (p < ((k - m) shl 1)) {
          for (i in 0 until m) {
            compSwap(arr, j + p + i, j + p + m + i, end)
          }
          p += m shl 1
        }
        j += k shl 1
      }
      m = m shr 1
    }
    k = k shl 1
  }
}

fun main() {
  var array = arrayOf<Int>(
    0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56,
  )
  sort(array)
  println("[%s]".format(array.joinToString(", ")))
}

File: App/Resources/AlgorithmDetails/pairwisemergesortiterative/pairwisemergesortiterative.go
package main

import (
	"fmt"
)

func compSwap(arr []int, a, b, end int) {
	if b < end && arr[a] > arr[b] {
		arr[a], arr[b] = arr[b], arr[a]
	}
}

func sort(arr []int) []int {
	length := len(arr)
	end := length

	n := 1
	for n < length {
		n <<= 1
	}

	k := n >> 1
	for k > 0 {
		j := 0
		for j < length {
			for i := 0; i < k; i++ {
				compSwap(arr, j+i, j+k+i, end)
			}
			j += k << 1
		}
		k >>= 1
	}

	k = 2
	for k < n {
		m := k >> 1
		for m > 0 {
			j := 0
			for j < length {
				p := m
				for p < ((k - m) << 1) {
					for i := 0; i < m; i++ {
						compSwap(arr, j+p+i, j+p+m+i, end)
					}
					p += m << 1
				}
				j += k << 1
			}
			m >>= 1
		}
		k <<= 1
	}
	return arr
}

func main() {
	array := []int{0, 39, 21, 62, 91, 77, 14, 23,
		90, 69, 51, 81, 68, 83, 32, 56}
	fmt.Println(sort(array))
}
